src/components/AugmentedResultList/AugmentedResultList.ts

```diff
@@ -0,0 +1,135 @@
+     * Creates a new `AugmentedResultList` component.
+     * @param element The HTMLElement on which to instantiate the component.
+     * @param options The options for the `ResultList` component.
+     * @param bindings The bindings that the component requires to function normally. If not set, these will be
+     * automatically resolved (with a slower execution time).
+     */
+    constructor(public element: HTMLElement, public options: AugmentedResultListOptions, public bindings: IComponentBindings) {
+        super(element, ComponentOptions.initComponentOptions(element, AugmentedResultList, options), bindings, AugmentedResultList.ID);
+        this.options.matchingFunction = this.options.matchingFunction ?? this.defaultMatchingFunction;
+    }
+
+    private defaultMatchingFunction = (augmentData: any, queryResult: IQueryResult) => {
+        const fieldName = this.getMatchingFieldString();
+        return augmentData[fieldName] === queryResult.raw[fieldName];
+    };
+
+    private getObjectPayload(results: IQueryResult[]): String[] {
+        const field = this.getMatchingFieldString();
+        return results.filter((result) => result.raw && result.raw[field]).map((result) => result.raw[field]);
+    }
+
+    private getMatchingFieldString() {
+        return this.options.matchingIdField.replace('@', '');
+    }
+
+    public renderResults(resultElements: HTMLElement[], append = false): Promise<void> {
+        return super.renderResults(resultElements, append);
+    }
+
+    public async buildResults(queryResults: Coveo.IQueryResults): Promise<HTMLElement[]> {
+        let remoteResults: IPromiseReturnArgs<IAugmentData>;
+        const fieldName = this.getMatchingFieldString();
+
+        if (this.options.fetchAugmentData) {
+            try {
+                // Call remote action to fetch augmenting data
+                remoteResults = await this.options.fetchAugmentData(this.getObjectPayload(queryResults.results));
+            } catch (e) {
+                this.logger.error(['Unable to fetch augment data.', e]);
+                return null;
+            }
+
+            if (remoteResults?.data) {
+                // Merge augmenting data with Coveo Results
+                queryResults.results.forEach((res: Coveo.IQueryResult) => {
+                    const match: any = remoteResults.data.resultData.find((data: any) => this.options.matchingFunction(data, res));
+
+                    // Attach data specific to each result/object
+                    for (const key in match) {
+                        if (key.toLowerCase() !== fieldName && Boolean(res.raw[key.toLowerCase()])) {
+                            this.logger.warn(`The ${key} field was overwritten on result: ${res.title}`);
+                        }
+                        res.raw[key.toLowerCase()] = match[key];
+                    }
+
+                    // Attach data common to all results
+                    for (const key in remoteResults.data.commonData) {
```

Comment:
I think we should merge the common data before the specific one. If we happen to have the same field in both common and specific data, I would expect the specific value to supersede the common one.